articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+For example, the `IntraProcessManager` has to take into account that potentially all the known `Subscription`s will take the message, regardless of their reliability QoS.
+If a `Publisher` or a `Subscription` are best-effort, they may not receive the meta-message thus preventing the `IntraProcessManager` from releasing the memory in the buffer.
+
+More details [here](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/).
+
+**TODO:** take into account also new QoS: Deadline, Liveliness and Lifespan
+[reference](https://github.com/ros2/design/pull/212).
+
+### Dependent on the RMW
+
+The current implementation of intra-process communication has to send meta-messages from the `Publisher` to the `Subscription`s.
+This is done using the `rmw_publish` function, the implementation of which depends on the chosen middleware.
+This results in the performance of a ROS2 application with intra-process communication enabled being heavily dependent on the chosen RMW implementation.
+
+Given the fact that these meta-messages have only to be received from entities within the same process, there is space for optimizing how they are transmitted by each RMW.
+However, at the moment none of the supported RMW is actively tackling this issue.
+This results in that the performance of a single process ROS2 application with intra-process communication enabled are still worst than what you could expect from a non-ROS application sharing memory between its components.
+
+In the following some experimental evidences are quickly presented.
+
+#### Memory requirement
+
+When a `Node` creates a `Publisher` or a `Subscription` to a topic `/MyTopic`, it will also create an additional one to the topic `/MyTopic/_intra`.
+The second topic is the one where meta-messages travel.
+Our [experimental results](https://github.com/irobot-ros/ros2-performance/tree/master/performances/experiments/crystal/pub_sub_memory#adding-more-nodes-x86_64) show that creating a `Publisher` or a `Subscription` has a non-negligible memory cost.
+This is particularly true for the default RMW implementation, Fast-RTPS, where the memory requirement increases almost expontentially with the number of participants and entities.
+
+#### Latency and CPU utilization
+
+Publishing a meta-message has the same overhead as that of publishing a small inter-process message.
+
+However, comparing the publication/reception of an intra and an inter-process message, the former requires several additional operations: it has to store the message in the ring buffer, monitor the number of `Subscription`s, and extract the message.
+
+The result is that from the latency and CPU utilization point of view, it is convenient to use intra-process communication only when the message size is at least 5KB.
+
+
+### Problems when both inter and intra-process communication are needed
+
+Currently, ROS2 does not provide any API for making nodes or `Publisher` and `Subscription` to ignore each other.
```

Comment:
A couple of lines below I specify that 
```
Note that these messages will be discarded, but they will still cause an overhead.
```

Do you think that I should describe this a little bit more? Maybe with a reference to the `ignore_local_publication` flag

Reply:
It's not necessarily true that when using `ignore_local_publication` that there is overhead in the middleware. Some DDS implementations will do publisher (datawriter) side filtering in this case, so no unused data is received.

Either way, I think the sentence being referenced here is factually inaccurate, because there is an API that deals with this in part, regardless of whether or not the way it is implemented is efficient or not (though depending on the middleware it may actually be quite efficient).